src/spmc_queue.ml

```diff
@@ -0,0 +1,244 @@
+      (* CAS here is not necessary but leaving it it for assertion.
+         Can be removed once we're confident. *)
+      assert (Atomic.compare_and_set head (old_head_val + 1) old_head_val);
+      (* Atomic.set head index; *)
+      None)
+    else
+      let index = old_head_val land mask in
+      let cell = Array.get array index in
+      let item = ref (Atomic.get cell) in
+      while
+        not (Option.is_some !item && Atomic.compare_and_set cell !item None)
+      do
+        item := Atomic.get cell
+      done;
+      !item
+
+  (* successfuly rolled back *)
+  let is_empty_thorough { head = _; tail; mask; array; _ } : bool =
+    let mask = Atomic.get mask in
+    let array = Atomic.get array in
+    let size = Array.length array in
+    let tail_value = Atomic.get tail in
+    let seen_not_free = ref false in
+    let i = ref (size - 1) in
+    while (not !seen_not_free) && !i >= 0 do
+      let cell = Array.get array ((tail_value + !i) land mask) in
+      seen_not_free := Atomic.get cell != None;
+      i := !i - 1
+    done;
+    not !seen_not_free
+
+  let is_empty queue =
+    let { head; tail; _ } = queue in
+    let tail_val = Atomic.get tail in
+    let head_val = Atomic.get head in
+    if tail_val - head_val != 0 then false else is_empty_thorough queue
+
+  let steal ~from ({ mask = local_mask; _ } as local) =
+    (* need to be careful with from queue, which is not local *)
+    let local_size = Atomic.get local_mask + 1 in
+    let ({
+           head = from_head;
+           tail = from_tail;
+           mask = from_mask;
+           array = from_array;
+           _;
+         }
+          : 'a t) =
+      from
+    in
+    let from_mask_val = Atomic.get from_mask in
+    let from_array = Atomic.get from_array in
+    (* assumes there's space in the queue *)
+    let from_tail_val = Atomic.get from_tail in
+    let from_head_val = Atomic.get from_head in
+    let from_size = from_tail_val - from_head_val in
+    if from_size < 1 then 0
+    else
+      let stealable =
+        (* steal even if there's a single element, thus *+1* *)
+        min ((from_size + 1) / 2) local_size
+      in
+      let new_from_head_val = from_head_val + stealable in
+      let acquired_items =
+        Atomic.compare_and_set from_head from_head_val new_from_head_val
+      in
+      assert (from_tail_val - new_from_head_val >= 0);
+      if not acquired_items then 0
+      else (
+        for i = 0 to stealable - 1 do
+          let from_index = (from_head_val + i) land from_mask_val in
+          let cell = Array.get from_array from_index in
+          let value = ref (Atomic.get cell) in
+          while
+            not
+              (Option.is_some !value && Atomic.compare_and_set cell !value None)
+          do
+            value := Atomic.get cell
+          done;
+          let value_exn = function None -> assert false | Some v -> v in
+          while not (push local (value_exn !value)) do
+            ()
+          done
```

Comment:
As in the other case, `stealer` has to ensure there's physical space in the queue. If it does not have such certainty, it needs to call `is_empty` before trying to steal.

We could add the index checks and array scan here, and make `steal` return `Not_enough_space if they fail. I'm weakly hesitant, since it's not the usual use-case, but it should fine to include.